presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java

```diff
@@ -0,0 +1,244 @@
+import io.prestosql.execution.StageState;
+import io.prestosql.execution.TaskInfo;
+import io.prestosql.execution.TaskManagerConfig;
+import io.prestosql.spi.QueryId;
+import io.prestosql.spi.connector.ColumnHandle;
+import io.prestosql.spi.predicate.Domain;
+import io.prestosql.spi.predicate.TupleDomain;
+import io.prestosql.sql.DynamicFilters;
+import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.tree.Expression;
+import io.prestosql.sql.tree.SymbolReference;
+
+import javax.annotation.PreDestroy;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.Immutable;
+import javax.annotation.concurrent.ThreadSafe;
+import javax.inject.Inject;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.stream.Collectors.groupingBy;
+import static java.util.stream.Collectors.mapping;
+import static java.util.stream.Collectors.toList;
+
+@ThreadSafe
+public class DynamicFilterService
+{
+    private final Map<SourceDescriptor, Optional<Domain>> dynamicFilterSummaries = new ConcurrentHashMap<>();
+
+    @GuardedBy("this")
+    private final Map<QueryId, Supplier<List<StageInfo>>> queries = new HashMap<>();
+
+    private final ScheduledExecutorService collectDynamicFiltersExecutor = newSingleThreadScheduledExecutor(daemonThreadsNamed("DynamicFilterService"));
+
+    @Inject
+    public DynamicFilterService(TaskManagerConfig taskConfig)
+    {
+        collectDynamicFiltersExecutor.scheduleWithFixedDelay(this::collectDynamicFilters, 0, taskConfig.getStatusRefreshMaxWait().toMillis(), MILLISECONDS);
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        collectDynamicFiltersExecutor.shutdownNow();
+    }
+
+    public void registerQuery(SqlQueryExecution sqlQueryExecution)
+    {
+        registerQuery(sqlQueryExecution.getQueryId(), sqlQueryExecution::getAllStages);
+    }
+
+    @VisibleForTesting
+    public synchronized void registerQuery(QueryId queryId, Supplier<List<StageInfo>> stageInfoSupplier)
+    {
+        queries.putIfAbsent(queryId, stageInfoSupplier);
+    }
+
+    @VisibleForTesting
+    public synchronized void collectDynamicFilters()
+    {
+        List<QueryId> queriesWithDynamicFilters = dynamicFilterSummaries.entrySet()
```

Comment:
Let's use `queries.keySet()`. I'm not sure `stageInfoSupplier` is that expensive that we need to optimize it's calls. Alternatively, we could populate `dynamicFilterSummaries` on `registerQuery` as we have access to plan using: `io.prestosql.execution.SqlQueryExecution#getQueryPlan`

Reply:
I've added code to populate `dynamicFilterSummaries` in `registerQuery` by using `PlanNodeSearcher` on the query plan,

Reply:
Using `queries.entrySet()` now